cita-executor/core/src/libexecutor/genesis.rs

```diff
@@ -99,83 +104,92 @@ impl Genesis {
+                    state
+                        .add_balance(&address, value)
+                        .expect("init balance fail");
+                }
+            }
+            for (key, values) in contract.storage.clone() {
+                state
+                    .set_storage(
+                        &address,
+                        H256::from_unaligned(key.as_ref()).unwrap(),
+                        H256::from_unaligned(values.as_ref()).unwrap(),
+                    )
+                    .expect("init code set_storage fail");
+            }
+        }
+        state.commit().expect("state commit error");
+        //query is store in chain
+        for (address, contract) in &self.spec.alloc {
+            let address = Address::from_unaligned(address.as_str()).unwrap();
+            for (key, values) in &contract.storage {
+                let result =
+                    state.get_storage(&address, &H256::from_unaligned(key.as_ref()).unwrap());
+                assert_eq!(
+                    H256::from_unaligned(values.as_ref()).unwrap(),
+                    result.expect("storage error")
+                );
+            }
+        }
+
+        trace!("**** end **** \n");
+        let root = state.root;
+        trace!("root {:?}", root);
+        self.block.set_state_root(root);
+        self.block.rehash();
+
+        self.save(state_db.database())
     }
 
-    // fn save(&mut self, state: State<StateDB>, db: &Arc<KeyValueDB>) -> Result<(), String> {
-    //     let mut batch = db.transaction();
-    //     let hash = self.block.hash().unwrap();
-    //     let height = self.block.number();
-    //     //初始化的时候需要获取头部信息
-    //     batch.write(db::COL_HEADERS, &hash, self.block.header());
-    //     batch.write(db::COL_EXTRA, &CurrentHash, &hash);
-    //     batch.write(db::COL_EXTRA, &height, &hash);
-    //     let mut state_db = state.drop().1;
-    //     state_db
-    //         .journal_under(&mut batch, height, &hash)
-    //         .expect("DB commit failed");
-    //     db.write(batch)
-    // }
+    fn save(&mut self, db: Arc<CitaDB>) -> Result<(), String> {
+        // FIXME: The decode for the value may not equal to origin one.
+        let hash = self.block.hash().unwrap().to_vec();
+        let height = self.block.number().to_be_bytes().to_vec();
+
+        let current_hash =
+            H256::from("7cabfb7709b29c16d9e876e876c9988d03f9c3414e1d3ff77ec1de2d0ee59f66");
+        // Need to get header in init function.
+        db.insert(
+            Some(DataCategory::Headers),
+            hash.clone(),
+            self.block.header().rlp(),
+        )
+        .expect("Insert block header error.");
+        db.insert(
+            Some(DataCategory::Extra),
+            current_hash.to_vec(),
+            hash.clone(),
+        )
+        .expect("Insert block hash error.");
+        db.insert(Some(DataCategory::Extra), height, hash.clone())
+            .expect("Insert block hash error.");
+
```

Comment:
Do not use `expect()` here, It will panic when error occured.  Catch the db errors and return please.

Reply:
I think can not insert a value to database is unacceptable， so I let it panic here.